### PR TITLE
Enable Validating Header Dependencies

### PR DESCRIPTION
**Pull request by envp**
*Issue number of the reported bug or feature request:  Internal ticket*

**Describe your changes**
We add the ability to track live dependencies of a header in the include graph through counts of:
* macro references
* variable references
* name re-declarations
* type references

provided by `file uid A` through `file uid B`. We call this the 'usage reference count' of the edge `{A, B}`.

Add new method:
  * `std::set<FileGraphEdge> IncludeGraph::liveDependencies(FileUID headerUID)`
    - Return the set of headers that provide symbols, macros, or types referenced by the source in the file with UID `headerUID`
  * `bool IncludeGraph::decrementRefCount(FileGraphEdge edge)`
    - For and `edge` in the include graph, decrement the usage reference count of this edge,
returning a boolean indicating the success of the operation. Operation is considered failed if the edge doesn't exist or if the count is already 0.

**Testing performed**
Added a test case `t/031-validate-incldue-graph.t.cpp` for the features proposed.

Test suite passes locally on:
- LLVM-8 / Darwin 18.6.0
- docker with `LLVM-7 / ubuntu:18.04`
- docker with `LLVM-8 / ubuntu:18.10`
- docker with `LLVM-9 / ubuntu:19.04`

### CODE REVIEW

**Comment by envp** [comment]
cc @dbeer1 @ruoso @azeemba

File: include/clangmetatool/collectors/include_graph.h [hunk 1]
```diff
@@ -46,6 +46,30 @@ class IncludeGraph {
...
+
+  /**
+   * Decrement the reference count tracked for this edge, notifying the caller
+   * if the operation was successful.
+   *
+   * Return a boolean if indicating we were able to decrement the count:
+   *  - false if the edge doesn't exist or has no balance to decrement
+   *  - true otherwise. This will always reduce the count by 1
+   */
+  bool decrementRefCount(const clangmetatool::types::FileGraphEdge &edge);
```
Comment (dbeer1) [inline include/clangmetatool/collectors/include_graph.h]: I don't think we should have functions here that operate on the data held by the collector. Instead, users can copy it out and modify it if needed. Maybe these functions should then be on the `IncludeGraphData` object itself?

Comment (envp) [reply include/clangmetatool/collectors/include_graph.h]: Agreed, will move this to be defined on the struct instead, or would a free function accepting a pointer be okay too?

File: include/clangmetatool/collectors/include_graph.h [hunk 2]
```diff
@@ -46,6 +46,20 @@ class IncludeGraph {
...
+   *
+   * Header B is called a 'live' dependency of B if A uses or references a
+   * name originating from B or any of its transitive dependencies, i.e.:
+   * - A references a macro provided through B
+   * - A references a variable declaration provided through B
+   * - A references a type declaration provided through B
+   * - A redeclares a named declaration in B.
+   */
+  std::set<types::FileUID>
+  liveDependencies(const clangmetatool::types::FileUID &headerFUID) const;
```
Comment (dbeer1) [inline include/clangmetatool/collectors/include_graph.h]: This should move out of the collector class too, right?

Comment (envp) [reply include/clangmetatool/collectors/include_graph.h]: Ack.

**Comment by envp** [comment]
@dbeer1 This is ready for review again.

**Comment by envp** [comment]
I've rebased the 'extra' commits into the intended parent.

File: include/clangmetatool/include_graph_dependencies.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#ifndef INCLUDED_CLANGMETATOOL_INCLUDE_GRAPH_DEPENDENCIES_H
+#define INCLUDED_CLANGMETATOOL_INCLUDE_GRAPH_DEPENDENCIES_H
+
+#include <clangmetatool/collectors/include_graph_data.h>
+
+namespace clangmetatool {
+struct IncludeGraphDependencies {
```
Comment (dbeer1) [inline include/clangmetatool/include_graph_dependencies.h]: Can you add doc here? Just something simple about how these functions can be used to manipulate and check dependencies in an include graph

Comment (envp) [reply include/clangmetatool/include_graph_dependencies.h]: Ack. Done.

**Comment by envp** [comment]
@azeemba This is good to merge now.

**Pull request merged**